lib/system/coroutines.nim

```diff
@@ -0,0 +1,125 @@
+## Internal module included by the system module. Implements the basic
+## compiler interface for coroutines, plus a low-level API for working
+## with them.
+
+type
+  CoroutineBase {.compilerproc.} = ref object of RootObj
```

Comment:
there seems to be a lack of transitions from child to parent
like `mom` field in CPS at least i dont get how coroutine can return control back to its parent after several suspendings without storing parent ref in self

Reply:
``` nim
proc a() {.coroutine.} =
  echo 2
  suspend()
  echo 3
proc b() {.coroutine.} =
  echo 1
  a()
  echo 4
b()
```
with current trampoline implementation this will work because there are not real suspends. i think it is more like closure iterators right now but with only single return at the end (so even less usefull)

but if we save a coroutine somewhere and will try to continue later it will lost its parent because it lost its stack on which it all was worked before

Reply:
> there seems to be a lack of transitions from child to parent like `mom` field in CPS

Correct, at the moment, the basic instance type has no concept of coroutine instance relationship.

> at least i dont get how coroutine can return control back to its parent after several suspendings without storing parent ref in self

It is currently possible to implement tail calls, which I think is what you're after, by using a custom instance type that stores the parent instance. If knowing the parent coroutine instance (if any) is something that's necessary often enough, then I think it would make sense to have a `mom`-like field part of `Coroutine[T]`.

> with current trampoline implementation this will work because there are not real suspends

Yep, but keep in mind that the `trampoline` implementation is only a *showcase implementation*. It's there to have a working example of how an implementation *could* look like.

In addition, the `trampoline` is looked up in the context of `a()`, meaning that it's possible to override/overload the default `trampoline` (I'm still not sure whether there should be an always-present default `trampoline` at all).

Reply:
> > at least i dont get how coroutine can return control back to its parent after several suspendings without storing parent ref in self
> 
> It is currently possible to implement tail calls, which I think is what you're after, by using a custom instance type that stores the parent instance. If knowing the parent coroutine instance (if any) is something that's necessary often enough, then I think it would make sense to have a `mom`-like field part of `Coroutine[T]`.

Hmm, if the current implementation doesn't remember the parent, and the sorta unconventional control flow, at least based on coroutine "standards", would a potentially better name for this more stripped down concept be `suspendable`. It's a `suspendable` routine, as in it can suspend it's execution, but that's really the only guarantee it provides. It doesn't have the state (i.e.: `mom`) to honour returning control back to the parent.

I'm sorta borrowing terminology from Kotlin here. Don't mean for this to be out of the blue, been thinking about it for a while and this occurred to me just now. I might be putting too hard a requirement on how coroutines should behave, but given that subroutines are meant to be a special case of coroutine, I think the parent requirement seems reasonable (subroutines take control from and return control to the parent).